[cxx-interop] Import `typedef`-ed template instantiations (#32950)
swiftlang/swift: docs/ABI/Mangling.rst

underscore ``_``, the identifier for the string literal is prefixed with an
additional underscore ``_``.

Conventions for constructing names
----------------------------------

Swift is a language that integrates with multiple languages (C, C++, ObjC,
ObjC++). To achieve interoperability symbols from other languages have to be
adapted in order to be integrated into Swift.

Importing C and C++ structs
~~~~~~~~~~~~~~~~~~~

C and C++ structs use their original names and their owning modules. Since Swift
metadata symbols and value-witness-tables are not generated by the C compiler,
Swift generates them when C/C++ module is imported and stores them in the `__C`
module. This can be observed when mangling a Swift function that accepts a C/C++
struct as a parameter.

Importing C++ template instantiations behind typedef
~~~~~~~~~~~~~~~~~~~

The template instantiation is imported as a struct named `__CxxTemplateInst`
plus Itanium mangled type of the instantiation. Double underscore (denoting a
reserved C++ identifier) is used to discourage direct usage. The struct is
stored in the `__C` module, similarly to plain C++ structs.

> 

Author: Done.

.. code::

  sSo9CxxStructVMn // -> nominal type descriptor for __C.CxxStruct

Importing C++ class template instantiations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A class template instantiation is imported as a struct named
``__CxxTemplateInst`` plus Itanium mangled type of the instantiation (see the
``type`` production in the Itanium specification). Note that Itanium mangling is
used on all platforms, regardless of the ABI of the C++ toolchain, to ensure
that the mangled name is a valid Swift type name (this is not the case for MSVC
mangled names). A prefix with a double underscore (to ensure we have a reserved
C++ identifier) is added to limit the possibility for conflicts with names of
user-defined structs. The struct is notionally defined in the ``__C`` module,
similarly to regular C and C++ structs and classes. Consider the following C++
module:

.. code-block:: c++

  template<class T>
  struct MagicWrapper {
    T t;
  };

  struct MagicNumber {};

  typedef MagicWrapper<MagicNumber> WrappedMagicNumber;

> ``WrappedMagicNumber`` will be imported as a typealias for a struct
> ``__CxxTemplateInst12MagicWrapperI11MagicNumberE``. Interface of the imported
> module will look as follows:

Reviewer: ```suggestion
``WrappedMagicNumber`` is imported as a typealias for a struct
``__CxxTemplateInst12MagicWrapperI11MagicNumberE``. Interface of the imported
module looks as follows:
```
